shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -88,62 +137,62 @@ protected void encode(ChannelHandlerContext ctx, Object msg, ByteBuf out) throws
             Append append = (Append) msg;
             Session session = setupSegments.get(new SimpleImmutableEntry<>(append.segment, append.getWriterId()));
             validateAppend(append, session);
-            if (!append.segment.equals(segmentBeingAppendedTo) || !append.getWriterId().equals(writerIdPerformingAppends)) {
-                breakFromAppend(null, null, out);
-            }
             final ByteBuf data = append.getData().slice();
-            final int msgSize = data.readableBytes();
             final AppendBatchSizeTracker blockSizeSupplier = (appendTracker == null) ? null :
                     appendTracker.apply(append.getFlowId());
 
-            session.lastEventNumber = append.getEventNumber();
-            session.eventCount++;
             if (blockSizeSupplier != null) {
-                blockSizeSupplier.recordAppend(append.getEventNumber(), msgSize);
+                blockSizeSupplier.recordAppend(append.getEventNumber(), data.readableBytes());
             }
-            if (bytesLeftInBlock == 0) {
-                currentBlockSize = msgSize + TYPE_PLUS_LENGTH_SIZE;
-                if (blockSizeSupplier != null) {
-                    currentBlockSize = Math.max(currentBlockSize, blockSizeSupplier.getAppendBlockSize());
-                }
-                bytesLeftInBlock = currentBlockSize;
-                segmentBeingAppendedTo = append.segment;
-                writerIdPerformingAppends = append.writerId;
-                writeMessage(new AppendBlock(session.id), out);
-                if (ctx != null && currentBlockSize > (msgSize + TYPE_PLUS_LENGTH_SIZE)) {
-                    ctx.executor().schedule(new BlockTimeouter(ctx.channel(), tokenCounter.incrementAndGet()),
-                                            blockSizeSupplier.getBatchTimeout(),
-                                            TimeUnit.MILLISECONDS);
-                }
-            }
-            // Is there enough space for a subsequent message after this one?
-            if (bytesLeftInBlock - msgSize > TYPE_PLUS_LENGTH_SIZE) {
-                out.writeBytes(data);
-                bytesLeftInBlock -= msgSize;
+
+            if (isChannelFree()) {
+                session.flush(out);
```

Comment:
Is this needed? I am confused why this can be immediately before the start append below.

Reply:
yes.. in case if the append completed (not break) , then channel become free and before it takes starts new session , the existing session data need be flushed.. there random failures such as Data already appended were observed without this.

Reply:
I am confused by this. If there is data on the session, and then the channel becomes free. it could flush all of the sessions at the time it is freeing the channel. If it does not, then why would the session need to flush before new data is written to the same session? It could simply add the new data to the session and then flush thus reducing the number of messages going over the wire.

Reply:
yes, i have done this change now.